src/lib/support/CHIPSparseArray.h

```diff
@@ -0,0 +1,193 @@
+#include <support/Base64.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/SafeInt.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+template <uint16_t Capacity>
+class SparseU64Array
+{
+
+public:
+    SparseU64Array() { CHIP_ZERO_AT(mArray); }
+    /**
+     * @brief
+     *   Serialize the sparse array into a base 64 encoded string.
+     *   Only the values till mNextAvailable index are encoded.
+     *   The empty indexes between 0, and mNextAvailable, are also
+     *   encoded.
+     *
+     * @param[in] buf Buffer where serialized string is written
+     * @param[in] buflen Length of buf
+     * @return pointer to buf, or nullptr in case of error
+     */
+    const char * Serialize(char * buf, uint16_t & buflen)
+    {
+        char * out = nullptr;
+        size_t len = sizeof(uint64_t) * mNextAvailable;
+        VerifyOrExit(buflen >= SerializedSize(), buflen = SerializedSize());
+        VerifyOrExit(buf != nullptr, buflen = SerializedSize());
+        VerifyOrExit(CanCastTo<uint16_t>(len), buflen = SerializedSize());
+
+        buflen = Base64Encode(reinterpret_cast<const uint8_t *>(mArray), static_cast<uint16_t>(len), buf);
+        out    = buf;
+
+    exit:
+        return out;
+    }
+
+    /**
+     * @brief
+     *   Deserialize a base64 encoded string into the sparse array.
+     *   The mNextAvailable index is calculated based on how many
+     *   values are in the deserialized array.
+     *
+     * @param[in] serialized Serialized buffer
+     * @param[in] buflen Length of buffer
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR Deserialize(const char * serialized, uint16_t buflen)
+    {
+        CHIP_ERROR err     = CHIP_NO_ERROR;
+        uint16_t decodelen = 0;
+        VerifyOrExit(buflen <= MaxSerializedSize(), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        decodelen = Base64Decode(serialized, buflen, reinterpret_cast<uint8_t *>(mArray));
+        VerifyOrExit(decodelen <= sizeof(uint64_t) * Capacity, err = CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrExit(decodelen % sizeof(uint64_t) == 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        mNextAvailable = decodelen / static_cast<uint16_t>(sizeof(uint64_t));
+    exit:
+        return err;
+    }
+
+    /**
+     * @brief
+     *   Get the length of string if the array is serialized.
+     */
+    uint16_t SerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mNextAvailable)); }
+
+    /**
+     * @brief
+     *   Get the maximum length of string if the array was full and serialized.
+     */
+    uint16_t MaxSerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * Capacity)); }
+
+    /**
+     * @brief
+     *   Find a value in the array.
+     *
+     * @param[in] value Value to find
+     * @return index of the value if found, or max length (Capacity) of the array
+     */
+    uint16_t Find(uint64_t value)
```

Comment:
is there a way to get a value at a given index, or is such an ask outside the scope of this facility?

Reply:
I was thinking of it being out of scope, since the index -> value mapping is not guaranteed.